### PR TITLE
Improve Windows Compatibility(for lib/TH)

### PR DESCRIPTION
**Pull request by peterjc123**
Win64 support for lib/TH

### CODE REVIEW

File: torch/lib/TH/THAtomic.c [hunk 1]
```diff
@@ -19,75 +19,75 @@
...
 
-int THAtomicAdd(int volatile *a, int value)
+int THAtomicAdd(int32_t volatile *a, int32_t value)
 {
 #if defined(USE_C11_ATOMICS)
   return atomic_fetch_add(a, value);
 #elif defined(USE_MSC_ATOMICS)
-  assert(sizeof(int) == sizeof(long));
-  return _InterlockedExchangeAdd((long*)a, value);
+  //assert(sizeof(int32_t) == sizeof(long));
```
Comment (fmassa) [inline torch/lib/TH/THAtomic.c]: we can probably remove this comment

File: torch/lib/TH/THAtomic.c [hunk 2]
```diff
@@ -19,75 +19,75 @@
...
 
-int THAtomicCompareAndSwap(int volatile *a, int oldvalue, int newvalue)
+int THAtomicCompareAndSwap(int32_t volatile *a, int32_t oldvalue, int32_t newvalue)
 {
 #if defined(USE_C11_ATOMICS)
   return atomic_compare_exchange_strong(a, &oldvalue, newvalue);
 #elif defined(USE_MSC_ATOMICS)
-  assert(sizeof(int) == sizeof(long));
-  return (_InterlockedCompareExchange((long*)a, (long)newvalue, (long)oldvalue) == (long)oldvalue);
+  //assert(sizeof(int32_t) == sizeof(long));
```
Comment (fmassa) [inline torch/lib/TH/THAtomic.c]: this one as well

File: torch/lib/TH/THFile.c [hunk 3]
```diff
@@ -147,11 +147,11 @@ IMPLEMENT_THFILE_SCALAR(Half, THHalf)
...
-IMPLEMENT_THFILE_STORAGE(Byte, unsigned char)
-IMPLEMENT_THFILE_STORAGE(Char, char)
-IMPLEMENT_THFILE_STORAGE(Short, short)
-IMPLEMENT_THFILE_STORAGE(Int, int)
-IMPLEMENT_THFILE_STORAGE(Long, long)
+IMPLEMENT_THFILE_STORAGE(Byte, uint8_t)
+IMPLEMENT_THFILE_STORAGE(Char, int8_t)
+IMPLEMENT_THFILE_STORAGE(Short, int16_t)
+IMPLEMENT_THFILE_STORAGE(Int, int32_t)
+IMPLEMENT_THFILE_STORAGE(Long, int32_t)
```
Comment (fmassa) [inline torch/lib/TH/THFile.c]: `int64_t`

File: torch/lib/TH/generic/THBlas.h [hunk 4]
```diff
@@ -3,17 +3,17 @@
 #else
 
 /* Level 1 */
-TH_API void THBlas_(swap)(long n, real *x, long incx, real *y, long incy);
-TH_API void THBlas_(scal)(long n, real a, real *x, long incx);
-TH_API void THBlas_(copy)(long n, real *x, long incx, real *y, long incy);
-TH_API void THBlas_(axpy)(long n, real a, real *x, long incx, real *y, long incy);
-TH_API real THBlas_(dot)(long n, real *x, long incx, real *y, long incy);
+TH_API void THBlas_(swap)(int64_t  n, real *x, int64_t  incx, real *y, int64_t  incy);
```
Comment (fmassa) [inline torch/lib/TH/generic/THBlas.h]: nit: there are two spaces after `int64_t` everywhere in this file

File: torch/lib/TH/generic/THTensor.c [hunk 5]
```diff
@@ -922,7 +922,7 @@ THDescBuff THTensor_(desc)(const THTensor *tensor) {
   int i;
   for(i = 0; i < tensor->nDimension; i++) {
     if(n >= L) break;
-    n += snprintf(str+n, L-n, "%ld", tensor->size[i]);
+    n += snprintf(str+n, L-n, "%lld", tensor->size[i]);
```
Comment (fmassa) [inline torch/lib/TH/generic/THTensor.c]: nit: is this the right format? I'm looking at https://stackoverflow.com/questions/9225567/how-to-print-a-int64-t-type-in-c and I'm not sure it's exactly what you are looking for. (This appears elsewhere in the codebase as well). Also, there are warning in the CI due to those changes, with warnings like
```
warning: format ‘%lld’ expects argument of type ‘long long int *’, but argument 3 has type ‘int64_t * {aka long int *}’ [-Wformat=]
```

File: torch/lib/TH/generic/THTensorRandom.c [hunk 6]
```diff
@@ -193,18 +193,19 @@ void THTensor_(multinomialAliasSetup)(THTensor *probs, THLongTensor *J, THTensor
...
+
+  for (i=0; i < output_nelem; i++)
     {
-      rand_ind = (long)THRandom_uniform(_generator, 0, K) ;
+      rand_ind = THRandom_uniform(_generator, 0, K);
+
       _q = THTensor_fastGet1d(q, rand_ind);
 
-      _mask = THRandom_bernoulli(_generator, _q);
+      _mask = (int) THRandom_bernoulli(_generator, _q);
```
Comment (fmassa) [inline torch/lib/TH/generic/THTensorRandom.c]: `_mask` is a `int64_t`, so no need for the casting I think

File: torch/lib/TH/THDiskFile.c [hunk 7]
```diff
@@ -556,7 +556,7 @@ static size_t THDiskFile_readString(THFile *self, const char *format, char **str
         total += TBRS_BSZ;
         p = THRealloc(p, total);
       }
-      if (fgets(p+pos, total-pos, dfself->handle) == NULL) /* eof? */
+      if (fgets(p+pos, (int) (total-pos), dfself->handle) == NULL) /* eof? */
       {
         if(pos == 0)
```
Comment (fmassa) [inline torch/lib/TH/THDiskFile.c]: There are warnings in the contribuild down in this file, maybe you miss some castings? Example of warning
```
/home/travis/build/pytorch/pytorch/torch/lib/TH/THDiskFile.c: In function ‘THDiskFile_new’:
/home/travis/build/pytorch/pytorch/torch/lib/TH/THDiskFile.c:618:5: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
     THDiskFile_readChar,
     ^
```

**Review by fmassa — CHANGES_REQUESTED** [review]
Thanks for the PR!
I did a first pass on it, and had a look on the output of the contribuild. TH compiles, but the compiler outputs some warnings that might be good to fix before merging this PR.
Also, I didn't check the implementation inside `THAllocator.c`.

**Comment by peterjc123** [comment]
There 're no warnings now.

**Comment by fmassa** [comment]
Thanks for the changes!
TH seems now to compile without any warnings.
The build fails because of the dependency of THPP in TH, so we should probably only merge these PRs together.
@apaszke any better ideas on how we can run the tests with all the changes? The monolithic PR was too hard to review, the separate PRs are easier but running the full test suite is trickier.

**Comment by peterjc123** [comment]
@fmassa I tried to make CI builds with windows-full branch. Every time a change is made, it will be merged into this branch. Maybe it's useful.
For Linux/Mac    [Build Status](https://travis-ci.org/peterjc123/pytorch)
For Windows    [Build status](https://ci.appveyor.com/project/peterjc123/pytorch)

File: torch/lib/TH/generic/THTensorRandom.c [hunk 8]
```diff
@@ -5,15 +5,15 @@
...
+  TH_TENSOR_APPLY(real, self, *self_data = (int8_t)(THRandom_random(_generator) % (INT8_MAX+1)););
 #elif defined(TH_REAL_IS_SHORT)
-  TH_TENSOR_APPLY(real, self, *self_data = (short)(THRandom_random(_generator) % (SHRT_MAX+1)););
+  TH_TENSOR_APPLY(real, self, *self_data = (int16_t)(THRandom_random(_generator) % (INT16_MAX+1)););
 #elif defined(TH_REAL_IS_INT)
-  TH_TENSOR_APPLY(real, self, *self_data = (int)(THRandom_random(_generator) % (INT_MAX+1UL)););
+  TH_TENSOR_APPLY(real, self, *self_data = (int32_t)(THRandom_random(_generator) % (INT32_MAX+1UL)););
 #elif defined(TH_REAL_IS_LONG)
-  TH_TENSOR_APPLY(real, self, *self_data = (long)(THRandom_random(_generator) % (LONG_MAX+1UL)););
+  TH_TENSOR_APPLY(real, self, *self_data = (int64_t)THRandom_random(_generator););
```
Comment (fmassa) [inline torch/lib/TH/generic/THTensorRandom.c]: We don't need the modulo here I suppose?

**Review by fmassa — APPROVED** [review]
Thanks a lot for your effort!
I think this is ready for merge. TH builds without warnings.
Let's wait until the other folders are reviewed so that we can merge them all.

**Comment by soumith** [comment]
Thanks a ton @peterjc123 for doing this big PR. This is really good stuff.

**Pull request merged**